helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/ReadOnlyWagedRebalancer.java

```diff
@@ -48,6 +50,14 @@ public ReadOnlyWagedRebalancer(String metadataStoreAddress, String clusterName,
         ConstraintBasedAlgorithmFactory.getInstance(preferences), Optional.empty());
   }
 
+  @Override
+  protected List<HelixRebalanceException.Type> failureTypesToPropagate() {
+    // Also propagate FAILED_TO_CALCULATE for ReadOnlyWagedRebalancer
+    return Arrays.asList(HelixRebalanceException.Type.INVALID_REBALANCER_STATUS,
```

Comment:
nit, something like returning super.failureTypesToPropagate().add(HelixRebalanceException.Type.FAILED_TO_CALCULATE) ?

Reply:
I don't think that's better because ReadOnlyWagedRebalancer doesn't always propagate a superset of failure types; that might change in the future. Separate lists are clearer.

Reply:
I think ReadOnlyWagedRebalancer should cover all the failure types of the regular rebalancer. Otherwise, what to return if the basic rebalancer cannot finish the calculation? Based on this assumption, adding an additional item to the list would be easier to be maintained.

Reply:
I disagree with this: for example, some extension of `WagedRebalancer (if not `ReadOnlyWagedRebalancer`) could think UNKNOWN_FAILURE is okay and just take the last-known good result instead of an exception. 

I understand where you come from: if we add a new HelixRebalanceException.Type, it seems easier if we do super sets so we just modify the base class. However, even if we introduce a new type, each implementation should be reviewed to see if that type should be propagated. Imo this coupling is unnecessary here. If you don't feel strongly about this, I'd like to stick to separate lists.

Reply:
1. Let's worry about another extension of the WAGED rebalancer when they are created. And no matter what you do for the ReadOnlyWagedRebalancer, you can do it differently according to the new requirement later. Not relevant.
2. For ReadOnlyWagedRebalancer, I don't think we need to revisit. We can just carry over whatever WagedRebalancer has plus the FAILED_TO_CALCULATE failure type. If this is not guaranteed, then ReadOnlyWagedRebalancer cannot be used in the simulate tool, since it may cause different behavior. That's exactly why I strongly prefer we just call the method of super class here. It prevents potential bugs in the future.

Reply:
I can see your point about ReadOnlyWagedRebalancer now. Making changes accordingly.